src/Compilers/Core/Portable/MetadataReader/MetadataHelpers.cs

```diff
@@ -505,28 +505,53 @@ private static short InferTypeArityFromMetadataName(string emittedTypeName, out
                 return 0;
             }
 
-            // Given a name corresponding to <unmangledName>`<arity>,
-            // extract the arity.
-            string stringRepresentingArity = emittedTypeName.Substring(indexOfManglingChar);
-
-            int arity;
-            bool nonNumericCharFound = !int.TryParse(stringRepresentingArity, NumberStyles.None, CultureInfo.InvariantCulture, out arity);
-
-            if (nonNumericCharFound || arity < 0 || arity > short.MaxValue ||
-                stringRepresentingArity != arity.ToString())
+            // Given a name corresponding to <unmangledName>`<arity>, extract the arity.
+            if (!tryScanArity(emittedTypeName[indexOfManglingChar..], out var arity))
             {
                 suffixStartsAt = -1;
                 return 0;
             }
 
             suffixStartsAt = indexOfManglingChar - 1;
-            return (short)arity;
+            return arity;
+
+            static bool tryScanArity(ReadOnlySpan<char> aritySpan, out short arity)
+            {
+                // Common case: aritySpan is 1 digit between 0-9.  No need for any complex parsing logic in that case.
+                if (aritySpan is [var firstChar and >= '0' and <= '9'])
+                {
+                    arity = (short)(firstChar - '0');
+                    return true;
+                }
+
+#if NET
+                bool nonNumericCharFound = !int.TryParse(aritySpan, NumberStyles.None, CultureInfo.InvariantCulture, out var intArity);
+#else
+                bool nonNumericCharFound = !int.TryParse(aritySpan.ToString(), NumberStyles.None, CultureInfo.InvariantCulture, out var intArity);
```

Comment:
>bool nonNumericCharFound = !int.TryParse(aritySpan.ToString(), NumberStyles.None, CultureInfo.InvariantCulture, out var intArity); [](http://example.com/codeflow?start=16&length=130)

Perhaps we could accumulate the value character by character and avoid ```aritySpan.SequenceEqual(intArity.ToString().AsSpan())``` check below. I assume it is guarding against leading '0's. According to documentation, ```NumberStyles.None``` means "The string to be parsed must consist of integral decimal digits only".

Reply:
Also, we probably can dismiss arities longer than 5 characters right away.

Reply:
It looks like there was no response to this thread.

Reply:
by and large, i felt that keeping things aligned here was safest.  actually going and changing logic felt risky, and it is something i'd prefer to do outside of hte PR.  The check of `1-9` felt very safe, and at least helped mitigate an allocation for the netfx case in the most common (presumably 99.99% generic case).  So expending effort on what is likely the 0.01% case just didn't seem worth it.

Reply:
Feel free